README.md
# App custom payment

[![Deploy](https://github.com/ecomplus/app-custom-payment/workflows/Deploy/badge.svg)](https://github.com/ecomplus/app-custom-payment/actions?workflow=Deploy)  [![Deploy](https://github.com/ecomplus/app-custom-payment/workflows/Publish/badge.svg)](https://github.com/ecomplus/app-custom-payment/actions?workflow=Publish) [![License MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

E-Com Plus app for custom payment methods

[CHANGELOG](https://github.com/ecomplus/app-custom-payment/blob/master/CHANGELOG.md)

# E-Com Plus Application Starter

:dog: Boilerplate for E-Com Plus apps with [Firebase](https://firebase.google.com/) Cloud Functions and GitHub Actions.

## Getting started

1. Start creating a [Firebase project](https://console.firebase.google.com/):
    - Analytics is not needed;
    - Set a nice project name (ID) and remember it;

2. Enter the project, go to _databases_ page (on menu) and _create database_:
    - Just bypass with default production mode and rules;

3. Get your Firebase token from CLI:
```bash
npm install -g firebase-tools
firebase login:ci
```

4. [Use this template](https://github.com/ecomplus/application-starter/generate) to generate a new repository for your application;

5. Go to your repository _settings_ tab and set the following _secrets_:
    - `FIREBASE_PROJECT_ID`: The ID (name) of your Firebase project;
    - `FIREBASE_TOKEN`: The token generated with `firebase-tools`;
    - `SERVER_OPERATOR_TOKEN`: Random (at least 16 bytes) admin token generated from CLI or [here](https://randomkeygen.com/);
    - `MARKET_TOKEN` (optional): Your E-Com Plus Market [personal token](https://market.e-com.plus/partners/account) to publish the app automatically;

## Next steps

Almost ready, time to :coffee: and code!

Edit [`functions/ecom.config.js`](functions/ecom.config.js) to set correct `app_id`, `title` and optionally more fields on base app body.

If you're not yet familiarized with this boilerplate, **read with attention the comments and instructions at the configuration file**. You can also setup procedures from there to specify the web-hooks your app should receive.

### E-Com common routes

All endpoints inside [`/ecom/`](functions/routes/ecom) path **MUST BE PRIVATE** (trey are by default), accessible only for E-Com Plus official servers or with `X-Operator-Token` header (equal to `SERVER_OPERATOR_TOKEN` env).

Requests to E-Com routes must have `X-Store-ID` header or `store_id` parameter on URL query string.

#### [`auth-callback`](functions/routes/ecom/auth-callback.js)

Handle [authentication callback request](https://developers.e-com.plus/docs/api/#/store/authenticate-app/authenticate-app) and save tokens to database, also create Store API procedures if configured.

> You can edit it to send custom requests to external server notifying new store installation or setup.

#### [`get-auth`](functions/routes/ecom/get-auth.js)

Returns Store API authentication data based on received Store ID, for external usage.

Sample response:

```json
{
  "application_id": "a00000000000000000000012",
  "application_app_id": 9000,
  "store_id": 1011,
  "authentication_id": "120000000000000000000012",
  "access_token": "eyJhbGciOi.eyYzdWIi.ZEONFh7HgQ"
}
```

> You should consume it if you want to use this boilerplate as an _authentication backend_ only. By getting this data (HTTP GET) you'll be able to run authenticated requests to [Store API](https://developers.e-com.plus/docs/api/#/store/) from your own server, with your preferred language, framework...

#### [`refresh-tokens`](functions/routes/ecom/refresh-tokens.js)

Call update service to start refreshing old access tokens.

#### [`webhook`](functions/routes/ecom/webhook.js)

Receives E-Com Plus notification POSTs for configured procedures.

> You may want to edit it to properly treat trigger body and do your custom stuff for each notification type.

#### [`modules/*`](functions/routes/ecom/modules)

Endpoints for [E-Com Plus Mods API](https://developers.e-com.plus/modules-api/).

> You should edit the respective ones for enabled modules (on `ecom.config.js`).

### Additional Store API handlers

If the app works with `admin_settings` (configuration), use [`getAppData`](functions/lib/store-api/get-app-data.js) to get the application `data`/`hidden_data` configured by merchant from dashboard through Store API ([Application Object](https://developers.e-com.plus/docs/api/#/store/applications/)).

We recommend treating Store API clients (40x) and server (50x) errors with [`errorHandling`](functions/lib/store-api/error-handling.js) abstraction.

### Application SDK

You may want to use [`appSdk`](https://github.com/ecomplus/application-sdk) to make custom authenticated requests to [Store API](https://developers.e-com.plus/docs/api/#/store/) or deeper coding out of our examples.

### Custom integration scripts

- Add third party libraries to [`functions/package.json`](functions/package.json) dependencies;
- Add custom web app routes by creating new files to [`functions/routes`](functions/routes) folder;
- Add new methods/handlers at [`functions/lib`](functions/lib) folder;

### Optional setup for [Market](https://market.e-com.plus/)

You may want to make your app public and available to install within E-Com Plus dashboard, to do this:

- Edit [`ecomplus-market.json`](ecomplus-market.json) to set a short description and category between _shipping_, _sales_, _tools_, _customer-service_, _marketing_, _inventory_ or _product-sourcing_;
- Overwrite the default PNG icon at [`functions/public/icon.png`](functions/public/icon.png);
- Write a good markdown description for the app at [`functions/public/description.md`](functions/public/description.md);
- Take note about [publication](#publication);

> If you are not yet a registered partner, start by filling out [this form](https://docs.google.com/forms/d/e/1FAIpQLSfd8uUsMG6N_rSFi2blGuk3Rfqi_BPp6fxschkmkdhEBVDsyw/viewform) and wait a little for our response :handshake:

## Examples

:sunglasses: [Awesome E-Com Plus applications](https://github.com/ecomplus/awesome#applications) to get inspired!

## Continuous integration

Every commit will trigger a new **deploy** (with [GitHub Actions](.github/workflows)), then your app will be accessible at:

`https://us-central1-<project-id>.cloudfunctions.net/app/` :blush:

The `assets/application.json` will be updated automatically with some package info and current Cloud Function endpoints, you can use it as body to [_Create new Application_](https://developers.e-com.plus/docs/api/#/store/applications/new-application) on Store API.

> You can skip deploy workflow by adding `[skip ci]` to the commit message.

Also, your app's access tokens to Store API will be **automatically refreshed** every 8 hours by scheduled workflow.

### Publication

If `MARKET_TOKEN` secret is set, the app will be **automatically published to [Market](https://market.e-com.plus/)** every time `assets/application.json` is changed, then merchants will be able to install it with up to date version.

## Developing and testing locally

Setup the project normally by cloning your repository and installing npm dependencies:

```bash
git clone git@github.com:$username/$app_repository.git
cd $app_repository
npm i
cd functions
npm i
```

Then you can call `npm run deploy` locally getting configuration from `.env` file on project root. You can also set a different `FIREBASE_PROJECT_ID` locally for test purpose only.

When you're done, you can publish the app by running `npm run publish:market`.

### Firebase tools

You can also use [`firebase-tools` CLI](https://firebase.google.com/docs/cli) to run tests/deploy with custom config or scripts.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.4.2](https://github.com/ecomplus/application-starter/compare/v1.4.1...v1.4.2) (2021-03-12)


### Bug Fixes

* **create-transaction:** prevent undefined `paymentOption` when no method name set ([e3428ee](https://github.com/ecomplus/application-starter/commit/e3428ee4e21f578504eca8c62d888e73e354aea4))
* **deps:** update all non-major dependencies ([#39](https://github.com/ecomplus/application-starter/issues/39)) ([f9b65fe](https://github.com/ecomplus/application-starter/commit/f9b65fed589744c942f436fffeabf78d1a89025f))

### [1.4.1](https://github.com/ecomplus/application-starter/compare/v1.4.0...v1.4.1) (2021-02-14)


### Bug Fixes

* **deps:** update all non-major dependencies ([#35](https://github.com/ecomplus/application-starter/issues/35)) ([e95ca30](https://github.com/ecomplus/application-starter/commit/e95ca30305c9aa719d6016aca44486ddf89e8641))
* **list-payments:** set configured  on  object ([4c5a10f](https://github.com/ecomplus/application-starter/commit/4c5a10f26109a1752638c16cd1d896b72a553d20))

## [1.4.0](https://github.com/ecomplus/application-starter/compare/v1.3.1...v1.4.0) (2021-02-05)


### Features

* **create-transaction:** handle configured payment link and redirect to payment ([cffc072](https://github.com/ecomplus/application-starter/commit/cffc072a490b052ce825f78b4bdde38d01875f98))
* **ecom-config:** add `payment_link` field to admin settings payment option schema ([b4164f6](https://github.com/ecomplus/application-starter/commit/b4164f6c86ab9f1ab8784d042d484413639839b6))


### Bug Fixes

* **ecom-config:** unset app authentication ([ba57513](https://github.com/ecomplus/application-starter/commit/ba575134ab378e42e0042f171b0542f1a6d4e912))
* **list-payments:** prevent error with undefined `paymentOption.payment_method` ([487f16a](https://github.com/ecomplus/application-starter/commit/487f16adecebf89437d178c433276b0263df3e1a))

### [1.3.1](https://github.com/ecomplus/application-starter/compare/v1.3.0...v1.3.1) (2020-12-17)


### Bug Fixes

* **deps:** update all non-major dependencies ([#27](https://github.com/ecomplus/application-starter/issues/27)) ([b94339e](https://github.com/ecomplus/application-starter/commit/b94339ed15b2cadfcde1871918bd1c872c288f2c))
* **deps:** update all non-major dependencies ([#31](https://github.com/ecomplus/application-starter/issues/31)) ([ca86478](https://github.com/ecomplus/application-starter/commit/ca86478a36be26d1efc55242ff17dc6fe0d25aef))
* **list-payments:** dont check cumulative discount when 'amount.discount' undefined ([#32](https://github.com/ecomplus/application-starter/issues/32)) ([eb99d1f](https://github.com/ecomplus/application-starter/commit/eb99d1ffb2117af17849d3a24ab58cade257e9fc))

## [1.3.0](https://github.com/ecomplus/application-starter/compare/v1.0.1...v1.3.0) (2020-12-11)


### Features

* **admin-settings:** add 'zip_range' to payment options config ([7967f9d](https://github.com/ecomplus/application-starter/commit/7967f9d30151984b78f2080a1680d2e20bf06671))
* **discount:** add option for non-cumulative discount ([#20](https://github.com/ecomplus/application-starter/issues/20)) ([cf4a92e](https://github.com/ecomplus/application-starter/commit/cf4a92e7c650dcff6a1d37c5a225ccada8a29aaf))
* **list-payment:** optionally handling address specific payment option ([3fa172e](https://github.com/ecomplus/application-starter/commit/3fa172e042e93c17c09fb7aaeb966376be8bf6b5))


### Bug Fixes

* **apply-discounts:** fix setting discount option and checking cumulative ([#19](https://github.com/ecomplus/application-starter/issues/19)) ([ee9b7b6](https://github.com/ecomplus/application-starter/commit/ee9b7b63b8352619d2cd5d4dce85981209c3e1ab))
* **deps:** update @ecomplus/application-sdk and firebase-admin major ([f46590f](https://github.com/ecomplus/application-starter/commit/f46590fca10a02869bfdb3e7e6a03604ec2a665e))
* **deps:** update @ecomplus/application-sdk firestore verison ([1c2d4ee](https://github.com/ecomplus/application-starter/commit/1c2d4ee570fd34289647d78880439dfbfcca07e8))
* **deps:** update all non-major dependencies ([#10](https://github.com/ecomplus/application-starter/issues/10)) ([2954b8a](https://github.com/ecomplus/application-starter/commit/2954b8aad62983e60359776aace955154cb07ac6))
* **deps:** update all non-major dependencies ([#23](https://github.com/ecomplus/application-starter/issues/23)) ([ad757cf](https://github.com/ecomplus/application-starter/commit/ad757cff78b250559f65535f8bc91d4505565f27))
* **list-payments:** fix pushing payment gateway objects ([466db39](https://github.com/ecomplus/application-starter/commit/466db398de068145896ffdefd0dc1764d3b18b5d))
* **list-payments:** payment method label for default discount option ([e7084db](https://github.com/ecomplus/application-starter/commit/e7084db854eb6bcb7195b3b09581b2d4e52d3869))
* **list-payments:** relaxing zip condition filter ([8d837fd](https://github.com/ecomplus/application-starter/commit/8d837fd87bfac444445633700f515d9e6e6ecd1f))
* **list-payments.js:** check if the discount is greater than 0 ([07696b4](https://github.com/ecomplus/application-starter/commit/07696b4055949d5471059e6eef03b661366e2ffa))
* **refresh-tokens:** add scheduled cloud function to run update ([322f882](https://github.com/ecomplus/application-starter/commit/322f882a13639346fbace66ebb3c465de9d8025d))
* **setup:** do not disable updates on refresh-tokens route ([a388827](https://github.com/ecomplus/application-starter/commit/a3888273958ae09136f2af9923be5f6771c6f9e8))
* cumulative_discount, amount.discount ([2d74ece](https://github.com/ecomplus/application-starter/commit/2d74ece737308d0380c56e686e0d28463af65ee2))
* description text ([d5fd4f9](https://github.com/ecomplus/application-starter/commit/d5fd4f9f31f5c3164d83267256296ec98db392d3))


functions/index.js
#!/usr/bin/env node

const { functionName, operatorToken } = require('./__env')

const path = require('path')
const recursiveReadDir = require('./lib/recursive-read-dir')

// Firebase SDKs to setup cloud functions and access Firestore database
const admin = require('firebase-admin')
const functions = require('firebase-functions')
admin.initializeApp()

// web server with Express
const express = require('express')
const bodyParser = require('body-parser')
const server = express()
const router = express.Router()
const routes = './routes'

// enable/disable some E-Com common routes based on configuration
const { app, procedures } = require('./ecom.config')

// handle app authentication to Store API
// https://github.com/ecomplus/application-sdk
const { ecomServerIps, setup } = require('@ecomplus/application-sdk')

server.use(bodyParser.urlencoded({ extended: false }))
server.use(bodyParser.json())

server.use((req, res, next) => {
  if (req.url.startsWith('/ecom/')) {
    // get E-Com Plus Store ID from request header
    req.storeId = parseInt(req.get('x-store-id') || req.query.store_id, 10)
    if (req.url.startsWith('/ecom/modules/')) {
      // request from Mods API
      // https://github.com/ecomclub/modules-api
      const { body } = req
      if (typeof body !== 'object' || body === null || !body.params || !body.application) {
        return res.status(406).send('Request not comming from Mods API? Invalid body')
      }
    }

    if (process.env.NODE_ENV !== 'development') {
      // check for operator token
      if (operatorToken !== req.get('x-operator-token')) {
        // last check for IP address from E-Com Plus servers
        const clientIp = req.get('x-forwarded-for') || req.connection.remoteAddress
        if (process.env.NODE_ENV !== 'development' && ecomServerIps.indexOf(clientIp) === -1) {
          return res.status(403).send('Who are you? Unauthorized IP address')
        }
      }
    }
  }

  // pass to the endpoint handler
  // next Express middleware
  next()
})

router.get('/', (req, res) => {
  // pretty print application body
  server.set('json spaces', 2)
  require(`${routes}/`)(req, res)
})

const prepareAppSdk = () => {
  // debug ecomAuth processes and ensure enable token updates by default
  process.env.ECOM_AUTH_DEBUG = 'true'
  process.env.ECOM_AUTH_UPDATE = 'enabled'
  // setup ecomAuth client with Firestore instance
  return setup(null, true, admin.firestore())
}

// base routes for E-Com Plus Store API
const routesDir = path.join(__dirname, routes)
recursiveReadDir(routesDir).filter(filepath => filepath.endsWith('.js')).forEach(filepath => {
  // set filename eg.: '/ecom/auth-callback'
  let filename = filepath.replace(routesDir, '').replace(/\.js$/i, '')
  if (path.sep !== '/') {
    filename = filename.split(path.sep).join('/')
  }
  if (filename.charAt(0) !== '/') {
    filename = `/${filename}`
  }

  // ignore some routes
  switch (filename) {
    case '/index':
      // home already set
      return
    case '/ecom/webhook':
      // don't need webhook endpoint if no procedures configured
      if (!procedures.length) {
        return
      }
      break
    default:
      if (filename.startsWith('/ecom/modules/')) {
        // check if module is enabled
        const modName = filename.split('/').pop().replace(/-/g, '_')
        if (!app.modules || !app.modules[modName] || app.modules[modName].enabled === false) {
          return
        }
      }
  }

  // expecting named exports with HTTP methods
  const methods = require(`${routes}${filename}`)
  for (const method in methods) {
    const middleware = methods[method]
    if (middleware) {
      router[method](filename, (req, res) => {
        console.log(`${method} ${filename}`)
        prepareAppSdk().then(appSdk => {
          middleware({ appSdk, admin }, req, res)
        }).catch(err => {
          console.error(err)
          res.status(500)
          res.send({
            error: 'SETUP',
            message: 'Can\'t setup `ecomAuth`, check Firebase console registers'
          })
        })
      })
    }
  }
})

server.use(router)
server.use(express.static('public'))

exports[functionName] = functions.https.onRequest(server)
console.log(`-- Starting '${app.title}' E-Com Plus app with Function '${functionName}'`)

// schedule update tokens job
// const cron = '10 */3 * * *'
/*
exports.updateTokens = functions.pubsub.schedule(cron).onRun(() => {
  return prepareAppSdk().then(appSdk => {
    return appSdk.updateTokens()
  })
})
console.log(`-- Sheduled update E-Com Plus tokens '${cron}'`)
*/


functions/routes/ecom/modules/list-payments.js
exports.post = ({ appSdk }, req, res) => {
  const { params, application } = req.body
  const amount = params.amount || {}
  const config = Object.assign({}, application.hidden_data, application.data)

  // start mounting response body
  // https://apx-mods.e-com.plus/api/v1/list_payments/response_schema.json?store_id=100
  const response = {
    payment_gateways: []
  }

  if (config.payment_options && config.payment_options.length) {
    config.payment_options.forEach(paymentOption => {
      if (paymentOption.zip_range) {
        // payment option for specific addresses only
        const { customer } = params
        if (customer && customer.addresses) {
          const address = customer.addresses.find(addr => addr.default) || customer.addresses[0]
          const zip = address && address.zip
          if (zip && (paymentOption.zip_range.min > zip || paymentOption.zip_range.max < zip)) {
            // zip code condition not satisfied
            return
          }
        }
      }

      const { label, icon, text, discount, enabled } = paymentOption
      if (enabled !== false) {
        const paymentMethod = paymentOption.payment_method
        if (!paymentMethod || !paymentMethod.code) {
          return
        }
        if (paymentOption.min_amount && (amount.total < paymentOption.min_amount)) {
          return
        }

        let paymentMethodName = paymentMethod.name || paymentMethod.code
        if (response.payment_gateways.find(gateway => {
          return gateway.payment_method && gateway.payment_method.name === paymentMethodName
        })) {
          paymentMethodName += ` ::${Math.floor(Math.random() * 1000)}`
        }

        const paymentGateway = {
          label,
          icon,
          text,
          payment_method: {
            code: paymentMethod.code,
            name: paymentMethodName
          },
          type: 'payment'
        }

        if (!amount.discount || paymentOption.cumulative_discount !== false) {
          paymentGateway.discount = discount
          if (discount && discount.value > 0) {
            // calculate discount value
            if (
              discount.apply_at !== 'freight' &&
              (!response.discount_option || (response.discount_option.value <= discount.value))
            ) {
              // default discount option
              response.discount_option = {
                label,
                ...discount
              }
            }
          }
        }

        response.payment_gateways.push(paymentGateway)
      }
    })
  }

  return res.send(response)
}


functions/routes/ecom/modules/create-transaction.js
exports.post = ({ appSdk }, req, res) => {
  const { params, application } = req.body
  const { amount, buyer } = params
  const config = Object.assign({}, application.data)

  let paymentMethodName = params.payment_method.name
  if (paymentMethodName) {
    paymentMethodName = paymentMethodName.split(' ::')[0]
  }

  let notes, paymentLink
  if (Array.isArray(config.payment_options) && config.payment_options.length) {
    let paymentOption = config.payment_options.find(paymentOption => {
      const paymentMethod = paymentOption.payment_method
      return paymentMethod &&
        paymentMethod.code === params.payment_method.code &&
        (!paymentMethod.name || paymentMethod.name === paymentMethodName)
    })
    if (!paymentOption) {
      paymentOption = config.payment_options[0]
    }
    if (typeof paymentOption.text === 'string') {
      notes = paymentOption.text.substr(0, 255)
    }
    if (typeof paymentOption.payment_link === 'string' && paymentOption.payment_link) {
      paymentLink = paymentOption.payment_link
        .replace(/{amount}/g, amount.total.toFixed(2).replace('.', ','))
        .replace(/{email}/g, buyer.email)
    }
  }

  const transaction = {
    payment_link: paymentLink,
    intermediator: {
      payment_method: params.payment_method
    },
    currency_id: params.currency_id,
    currency_symbol: params.currency_symbol,
    amount: amount.total,
    status: {
      current: 'pending'
    },
    flags: [
      'app-custom-payments'
    ],
    notes
  }

  const payload = {
    redirect_to_payment: Boolean(paymentLink),
    transaction
  }

  res.send(payload)
}


functions/ecom.config.js
/* eslint-disable comma-dangle, no-multi-spaces */

/**
 * Edit base E-Com Plus Application object here.
 * Ref.: https://developers.e-com.plus/docs/api/#/store/applications/
 */

const app = {
  app_id: 108091,
  title: 'Pagamento personalizado',
  slug: 'custom-payment',
  type: 'external',
  state: 'active',
  authentication: false,
  /**
   * Uncomment modules above to work with E-Com Plus Mods API on Storefront.
   * Ref.: https://developers.e-com.plus/modules-api/
   */
  modules: {
    /**
     * Triggered to calculate shipping options, must return values and deadlines.
     * Start editing `routes/ecom/modules/calculate-shipping.js`
     */
    // calculate_shipping:   { enabled: true },

    /**
     * Triggered to validate and apply discount value, must return discount and conditions.
     * Start editing `routes/ecom/modules/apply-discount.js`
     */
    // apply_discount:       { enabled: true },

    /**
     * Triggered when listing payments, must return available payment methods.
     * Start editing `routes/ecom/modules/list-payments.js`
     */
    list_payments: {
      enabled: true
    },

    /**
     * Triggered when order is being closed, must create payment transaction and return info.
     * Start editing `routes/ecom/modules/create-transaction.js`
     */
    create_transaction: {
      enabled: true
    }
  },

  /**
   * Uncomment only the resources/methods your app may need to consume through Store API.
   */
  auth_scope: {
    procedures: [
      //  'POST'           // Create procedures to receive webhooks
    ],
    products: [
      // 'GET',           // Read products with public and private fields
      // 'POST',          // Create products
      // 'PATCH',         // Edit products
      // 'PUT',           // Overwrite products
      // 'DELETE',        // Delete products
    ],
    brands: [
      // 'GET',           // List/read brands with public and private fields
      // 'POST',          // Create brands
      // 'PATCH',         // Edit brands
      // 'PUT',           // Overwrite brands
      // 'DELETE',        // Delete brands
    ],
    categories: [
      // 'GET',           // List/read categories with public and private fields
      // 'POST',          // Create categories
      // 'PATCH',         // Edit categories
      // 'PUT',           // Overwrite categories
      // 'DELETE',        // Delete categories
    ],
    customers: [
      // 'GET',           // List/read customers
      // 'POST',          // Create customers
      // 'PATCH',         // Edit customers
      // 'PUT',           // Overwrite customers
      // 'DELETE',        // Delete customers
    ],
    orders: [
      // 'GET',           // List/read orders with public and private fields
      // 'POST',          // Create orders
      // 'PATCH',         // Edit orders
      // 'PUT',           // Overwrite orders
      // 'DELETE',        // Delete orders
    ],
    carts: [
      // 'GET',           // List all carts (no auth needed to read specific cart only)
      // 'POST',          // Create carts
      // 'PATCH',         // Edit carts
      // 'PUT',           // Overwrite carts
      // 'DELETE',        // Delete carts
    ],

    /**
     * Prefer using 'fulfillments' and 'payment_history' subresources to manipulate update order status.
     */
    'orders/fulfillments': [
      // 'GET',           // List/read order fulfillment and tracking events
      // 'POST',          // Create fulfillment event with new status
      // 'DELETE',        // Delete fulfillment event
    ],
    'orders/payments_history': [
      // 'GET',           // List/read order payments history events
      // 'POST',          // Create payments history entry with new status
      // 'DELETE',        // Delete payments history entry
    ],

    /**
     * Set above 'quantity' and 'price' subresources if you don't need access for full product document.
     * Stock and price management only.
     */
    'products/quantity': [
      // 'GET',           // Read product available quantity
      // 'PUT',           // Set product stock quantity
    ],
    'products/variations/quantity': [
      // 'GET',           // Read variaton available quantity
      // 'PUT',           // Set variation stock quantity
    ],
    'products/price': [
      // 'GET',           // Read product current sale price
      // 'PUT',           // Set product sale price
    ],
    'products/variations/price': [
      // 'GET',           // Read variation current sale price
      // 'PUT',           // Set variation sale price
    ],

    /**
     * You can also set any other valid resource/subresource combination.
     * Ref.: https://developers.e-com.plus/docs/api/#/store/
     */
  },

  admin_settings: {
    payment_options: {
      schema: {
        type: 'array',
        title: 'Opções de pagamento customizadas',
        description: 'Configure e oferença novas opções que pagamento em sua loja',
        uniqueItems: true,
        items: {
          type: 'object',
          required: ['label', 'payment_method', 'enabled'],
          properties: {
            label: {
              type: 'string',
              maxLength: 50,
              title: 'Rótulo',
              description: 'Nome da forma de pagamento exibido para os clientes'
            },
            text: {
              type: 'string',
              maxLength: 1000,
              description: 'Texto auxiliar para forma que pagamento, pode-se usar tags HTML',
              title: 'Descrição'
            },
            icon: {
              type: 'string',
              maxLength: 255,
              format: 'uri',
              description: 'URL para ícone da forma de pagamento',
              title: 'Ícone'
            },
            payment_link: {
              type: 'string',
              maxLength: 255,
              format: 'uri',
              description: 'URL de redirecionamento para pagamento, pode-se usar as variáveis {amount} e {email}',
              title: 'Link de pagamento'
            },
            payment_method: {
              type: 'object',
              required: ['code'],
              additionalProperties: false,
              properties: {
                code: {
                  type: 'string',
                  enum: [
                    'credit_card',
                    'account_deposit',
                    'debit_card',
                    'balance_on_intermediary',
                    'loyalty_points',
                    'other'
                  ],
                  title: 'Código do método de pagamento'
                },
                name: {
                  type: 'string',
                  maxLength: 200,
                  title: 'Nome do método de pagamento',
                  description: 'Evite repetir o nome entre as opções de pagamento cadastradas'
                }
              },
              title: 'Método de pagamento'
            },
            discount: {
              type: 'object',
              required: ['value'],
              additionalProperties: false,
              properties: {
                apply_at: {
                  type: 'string',
                  enum: ['total', 'subtotal', 'freight'],
                  default: 'subtotal',
                  description: 'Em qual valor o desconto deverá ser aplicado no checkout'
                },
                type: {
                  type: 'string',
                  enum: ['percentage', 'fixed'],
                  default: 'percentage',
                  description: 'Valor percentual ou fixo a ser descontado, dependendo to tipo configurado'
                },
                value: {
                  type: 'number',
                  minimum: -99999999,
                  maximum: 99999999,
                  description: 'Valor do desconto'
                }
              },
              title: 'Desconto'
            },
            cumulative_discount: {
              type: 'boolean',
              default: true,
              title: 'Desconto cumulativo',
              description: 'Se a promoção poderá ser aplicada junto a cupons e campanhas de desconto'
            },
            min_amount: {
              type: 'integer',
              minimum: 1,
              maximum: 999999999,
              description: 'Valor mínimo para habilitar forma de pagamento',
              title: 'Valor mínimo'
            },
            zip_range: {
              title: 'Faixa de CEP',
              description: 'Limitar forma de pagamento por endereço do cliente',
              type: 'object',
              required: ['min', 'max'],
              properties: {
                min: {
                  type: 'integer',
                  minimum: 10000,
                  maximum: 999999999,
                  title: 'CEP inicial'
                },
                max: {
                  type: 'integer',
                  minimum: 10000,
                  maximum: 999999999,
                  title: 'CEP final'
                }
              }
            },
            enabled: {
              type: 'boolean',
              title: 'Habilitado',
              default: true
            }
          }
        }
      },
      hide: false
    }
  }
}

/**
 * List of Procedures to be created on each store after app installation.
 * Ref.: https://developers.e-com.plus/docs/api/#/store/procedures/
 */

const procedures = []

/**
 * Uncomment and edit code above to configure `triggers` and receive respective `webhooks`:

const { baseUri } = require('./__env')

procedures.push({
  title: app.title,

  triggers: [
    // Receive notifications when new order is created:
    {
      resource: 'orders',
      action: 'create',
    },

    // Receive notifications when order financial/fulfillment status changes:
    {
      resource: 'orders',
      field: 'financial_status',
    },
    {
      resource: 'orders',
      field: 'fulfillment_status',
    },

    // Receive notifications when products/variations stock quantity changes:
    {
      resource: 'products',
      field: 'quantity',
    },
    {
      resource: 'products',
      subresource: 'variations',
      field: 'quantity'
    },

    // Receive notifications when cart is edited:
    {
      resource: 'carts',
      action: 'change',
    },

    // Receive notifications when customer is deleted:
    {
      resource: 'customers',
      action: 'delete',
    },

    // Feel free to create custom combinations with any Store API resource, subresource, action and field.
  ],

  webhooks: [
    {
      api: {
        external_api: {
          uri: `${baseUri}/ecom/webhook`
        }
      },
      method: 'POST'
    }
  ]
})

 * You may also edit `routes/ecom/webhook.js` to treat notifications properly.
 */

exports.app = app

exports.procedures = procedures
